src/typechecker.erl

```diff
@@ -2909,116 +3025,154 @@ type_check_function(Env, {function,_, Name, NArgs, Clauses}) ->
-    ret(VEnv);
-add_types_pats([Pat | Pats], [Ty | Tys], TEnv, VEnv) ->
+%% Used for the arguments in clauses and the elements of tuples.
+%%
+%% The returned lists of types are interpreted like this:
+%% PatTy :: Pat as if Pat were a type. For match-all patterns, PatTy
+%% is the same as the type. For patterns matching a singleton type, PatTy
+%% is the singleton type. Otherwise, PatTy is none(). PatTy is a type exhausted
+%% by Pat. UBound is Ty or a subtype such that Pat :: UBound.
+add_types_pats(Pats, Tys, TEnv, VEnv) ->
+    add_types_pats(Pats, Tys, TEnv, VEnv, [], [], []).
+
+add_types_pats([], [], _TEnv, VEnv, PatTysAcc, UBoundsAcc, CsAcc) ->
+    {lists:reverse(PatTysAcc), lists:reverse(UBoundsAcc),
+     VEnv, constraints:combine(CsAcc)};
+add_types_pats([Pat | Pats], [Ty | Tys], TEnv, VEnv, PatTysAcc, UBoundsAcc, CsAcc) ->
     NormTy = normalize(Ty, TEnv),
-    {VEnv2, Cs1} = ?throw_orig_type(add_type_pat(Pat, NormTy, TEnv, VEnv),
-                                    Ty, NormTy),
-    {VEnv3, Cs2} = add_types_pats(Pats, Tys, TEnv, VEnv2),
-    {VEnv3, constraints:combine(Cs1, Cs2)}.
-
-%% Type check a pattern against a type and add variable bindings
--spec add_type_pat(Pat  :: any(), %% should be erl_parse:abstract_expr()
+    {PatTyNorm, UBoundNorm, VEnv2, Cs1} =
+        ?throw_orig_type(add_type_pat(Pat, NormTy, TEnv, VEnv),
+                         Ty, NormTy),
+    %% De-normalize the returned types if they are the type checked against.
+    PatTy  = case PatTyNorm  of NormTy -> Ty;
+                                _      -> PatTyNorm end,
+    UBound = case UBoundNorm of NormTy -> Ty;
+                                _      -> UBoundNorm end,
+    add_types_pats(Pats, Tys, TEnv, VEnv2,
+                   [PatTy|PatTysAcc], [UBound|UBoundsAcc], [Cs1|CsAcc]).
+
+%% Type check a pattern against a normalized type and add variable bindings.
+-spec add_type_pat(Pat  :: erl_parse:abstract_expr(),
                    Type :: type(),
                    TEnv :: #tenv{},
-                   VEnv :: #{}) ->
-          {NewVEnv :: #{}, constraints:constraints()}.
-add_type_pat({var, _, '_'}, _Ty, _TEnv, VEnv) ->
-    ret(VEnv);
-add_type_pat({var, _, A}, Ty, _TEnv, VEnv) ->
-    ret(VEnv#{ A => Ty });
-add_type_pat(Expr, {type, _, any, []}, _TEnv, VEnv) ->
-    ret(add_any_types_pat(Expr, VEnv));
+                   VEnv :: map()) ->
+          {PatTy :: type(), UBound :: type(), NewVEnv :: map(),
+           constraints:constraints()}.
+add_type_pat({var, _, '_'}, Ty, _TEnv, VEnv) ->
+    {Ty, Ty, VEnv, constraints:empty()};
+add_type_pat({var, P, A}=Var, Ty, TEnv, VEnv) ->
+    %% TODO: In a fun clause, A is always free, but not in e.g. case clauses
+    case VEnv of
+        #{A := VarTy} ->
+            case subtype(VarTy, Ty, TEnv) of
+                {true, Cs} ->
+                    {type(none), VarTy, VEnv, Cs};
+                false ->
+                    throw({type_error, tyVar, P, Var, VarTy, Ty})
```

Comment:
is it a type error in a case clause when Var :: a | b and Ty is a | c?
```
-spec f(a | b, a | c) -> boolean().
f(B,C) ->
  case C of B -> true
    _ -> false
  end.
```

Reply:
I think that makes sense. Compare
```erlang
-spec f(a | b, a | c) -> boolean().
f(B, C) ->
  case C of
    X when X == B -> true;  %% The operator '==' on line 31 requires arguments of compatible types.
    _ -> false              %% However, it has arguments of type a | c and a | b
  end.
```